keyring/backends/Windows/api.py

```diff
@@ -0,0 +1,273 @@
+
+        return result
+
+
+class CREDENTIALW(Structure):
+    _fields_ = [
+        ('Flags', DWORD),
+        ('Type', DWORD),
+        ('TargetName', c_wchar_p),
+        ('Comment', c_wchar_p),
+        ('LastWritten', FILETIME),
+        ('CredentialBlobSize', DWORD),
+        ('CredentialBlob', LPBYTE),
+        ('Persist', DWORD),
+        ('AttributeCount', DWORD),
+        ('Attributes', POINTER(CREDENTIAL_ATTRIBUTEW)),
+        ('TargetAlias', c_wchar_p),
+        ('UserName', c_wchar_p),
+    ]
+
+    def asdict(self):
+        result = {}
+
+        for field, _ in self._fields_:
+            value = getattr(self, field)
+
+            if field == 'LastWritten':
+                result[field] = value.asdict()
+            elif field == 'Attributes':
+                attr_dict = {}
+                for i in range(result['AttributeCount']):
+                    if value[i]:  # NULL pointer is treated as False by ctypes
+                        attr = value[i].asdict()
+                        attr_dict[attr['Keyword']] = attr
+                result[field] = attr_dict
+            elif field == 'CredentialBlob':
+                blob = string_at(value, getattr(self, 'CredentialBlobSize'))
+                result[field] = blob
+            else:
+                result[field] = value
+
+        return result
+
+
+# endregion
+
+
+# region Win32 API declarations
+
+
+def errcheck(result, func, args):
+    if not result:
+        raise WinError()
+
+
+_CredReadW = windll.advapi32.CredReadW
+_CredReadW.argtypes = [c_wchar_p, DWORD, DWORD, POINTER(POINTER(CREDENTIALW))]
+_CredReadW.restype = BOOL
+_CredReadW.errcheck = errcheck  # type: ignore
+
+_CredWriteW = windll.advapi32.CredWriteW
+_CredWriteW.argtypes = [POINTER(CREDENTIALW), DWORD]
+_CredWriteW.restype = BOOL
+_CredWriteW.errcheck = errcheck  # type: ignore
+
+_CredDeleteW = windll.advapi32.CredDeleteW
+_CredDeleteW.argtypes = [c_wchar_p, DWORD, DWORD]
+_CredDeleteW.restype = BOOL
+_CredDeleteW.errcheck = errcheck  # type: ignore
+
+_CredEnumerateW = windll.advapi32.CredEnumerateW
+_CredEnumerateW.argtypes = [
+    c_wchar_p,
+    DWORD,
+    POINTER(DWORD),
+    POINTER(POINTER(POINTER(CREDENTIALW))),
+]
+_CredEnumerateW.restype = BOOL
+_CredEnumerateW.errcheck = errcheck  # type: ignore
+
+_CredFree = windll.advapi32.CredFree
+_CredFree.argtypes = [LPVOID]
+_CredFree.restype = None
+
+CRED_TYPE_GENERIC = 1
+CRED_PERSIST_SESSION = 1
+CRED_PERSIST_LOCAL_MACHINE = 2
+CRED_PERSIST_ENTERPRISE = 3
+CRED_ENUMERATE_ALL_CREDENTIALS = 1
+CRED_MAX_VALUE_SIZE = 256
+CRED_MAX_ATTRIBUTES = 64
+MAX_PASSWORD_BYTES = CRED_MAX_VALUE_SIZE * CRED_MAX_ATTRIBUTES
+
+
+# endregion
+
+# region API
+
+
+def CredRead(Type, TargetName):
+    pcred = pointer(CREDENTIALW())
+    _CredReadW(TargetName, Type, 0, byref(pcred))
+    credential = pcred.contents.asdict()
+    _CredFree(pcred)
+    return credential
+
+
+def CredReadFromAttributes(Type, TargetName, Credential=None, encoding='utf-8'):
```

Comment:
nitpick: I'd probably name this using PEP 8 conventions. The reason for using the capital camel case is to match the names/parameters of the Windows API. In this case, the functionality is not part of the Windows API, but is a helper function, so probably should follow the naming of other helper functions.

Reply:
When I was writing it, I was trying to imagine the function I wish was already in the API, and wrote the one I wish was there.  But that's all inside my head.  I agree that it should be named as a helper, not as a faux member of the real API.